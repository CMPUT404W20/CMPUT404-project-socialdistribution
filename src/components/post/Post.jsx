import React, { Component } from "react";
import PropTypes from "prop-types";
import moment from "moment";
import Fade from "react-reveal/Fade";
import "../../styles/post/Post.scss";
import Dropdown from "react-bootstrap/Dropdown";
import DropdownButton from "react-bootstrap/DropdownButton";
import VisibilityOffIcon from "@material-ui/icons/VisibilityOff";
import TextareaAutosize from "react-textarea-autosize";
import ReactMarkdown from "react-markdown";
import breaks from "remark-breaks";
import Collapse from "react-bootstrap/Collapse";
import { CopyToClipboard } from "react-copy-to-clipboard";
import moreIcon from "../../images/more-icon.svg";
import * as CommentService from "../../services/CommentService";
import { userContext } from "../../contexts/UserContext";
import * as auth from "../../services/AuthenticationService";

class Post extends Component {
  constructor(props) {
    super(props);
    this.state = {
      commentSectionVisisble: false,
      newComment: "",
      currentUser: {},
    };
  }

  renderMenu = () => {
    const {
      post,
      invisible,
      previewMode,
      onEdit,
      onDelete,
    } = this.props;

    if (previewMode) {
      // add some gap at the top that would normally have been used by the menu bar
      // without this, the preview looks cramped
      return (<div className="spacer" />);
    }

    const dropdownIcon = <img id="post-more-icon" src={moreIcon} alt="more-icon" />;
    const formattedTime = moment(post.published).fromNow();
    const sharableLink = `${window.location.href}share/posts/${post.id}`;
    const postURL = new URL(post.source);
    const postHost = `${postURL.protocol}//${postURL.host}/`;
    return (
      <div className="post-info">
        <span className="post-user-and-visibility">
          {post.username}
          { invisible ? <VisibilityOffIcon fontSize="inherit" /> : null }
        </span>
        <userContext.Consumer>
          {(currentUser) => ((currentUser.host === postHost)
                && (
                <DropdownButton
                  id="post-more-button"
                  title={dropdownIcon}
                  drop="down"
                  alignRight
                >
                  <Fade left duration={500} distance="5px">
                    {/* the following enclosing tag is required for the fade to work properly */}
                    <>

                      {((currentUser.id === post.authorId)
                        ? (
                          <>
                            <Dropdown.Item onClick={() => onEdit(post.id)}>Edit</Dropdown.Item>
                            <Dropdown.Item onClick={() => onDelete(post.id)}>
                              Delete
                            </Dropdown.Item>
                          </>
                        )
                        : null)}

                      <CopyToClipboard text={sharableLink}>
                        <Dropdown.Item href="#">Copy Link</Dropdown.Item>
                      </CopyToClipboard>
                    </>
                  </Fade>
                </DropdownButton>
                )
          )}
        </userContext.Consumer>
        <div className="post-time">{formattedTime}</div>
      </div>
    );
  }

  renderComments = () => {
    const { post } = this.props;
    const comments = [];

    post.comments.forEach((comment) => {
      const opComment = comment.author.id === post.authorId;
      const formattedTime = moment(comment.published).fromNow();
      comments.push(
        <div key={comment.id}>
          <p>
            <span className={opComment ? "op" : ""}>{`${comment.author.displayName}:`}</span>
            <span className="comment-content">{comment.comment}</span>
          </p>
          <p className="comment-time">{formattedTime}</p>
        </div>,
      );
    });

    return comments;
  }

  handleCommentTextChange = (event) => {
    this.setState({ newComment: event.target.value });
  }

  handleSubmitNewComment = () => {
    const { post, onNewComment } = this.props;
    auth.getCurrentUser().then((user) => {
      this.setState({ currentUser: user.data }, () => {
        const { newComment, currentUser } = this.state;
        CommentService.createComment(post.source, post.id, newComment, currentUser)
          .then((success) => {
            if (success) {
            // clear the comment field but open the comment section
            // so the user can see the created post
              this.setState({
                newComment: "",
                commentSectionVisisble: true,
              });

              onNewComment();
            }
          }).catch((err) => {
            const error = err.response.data;
            // eslint-disable-next-line no-console
            console.log(error);
          });
      });
    });
  }

  handleCommentKeyPress = (event) => {
    if (event.key === "Enter") {
      event.preventDefault();
      this.handleSubmitNewComment();
    }
  }

  toggleCommentSection = () => {
    const { post } = this.props;
    if (post.comments.length > 0) {
      this.setState((prevState) => ({
        commentSectionVisisble: !prevState.commentSectionVisisble,
      }));
    }
  }

  renderCommentSection = () => {
    const { previewMode } = this.props;
    if (previewMode) {
      return null;
    }
    const { commentSectionVisisble, newComment } = this.state;
    const { post } = this.props;

    return (
      <div>
        <div className="button-block">
          <button
            className="post-show-comment"
            onClick={this.toggleCommentSection}
            aria-controls="post-comments"
            aria-expanded={commentSectionVisisble}
            type="button"
          >
            {post.comments.length}
            {" "}
            {post.comments.length === 1 ? "comment" : "comments"}
          </button>
        </div>
        <Collapse in={commentSectionVisisble}>
          {/* this div is necessary to prevent a choppy animation when opening the comments */}
          <div>
            <div className="comment-list">
              {this.renderComments()}
            </div>
          </div>
        </Collapse>
        <form className="make-comment-input-wrapper" action="submit" onSubmit={this.handleSubmitNewComment}>
          <TextareaAutosize
            placeholder="Add a comment"
            className="post-comment-text-area"
            onChange={this.handleCommentTextChange}
            onKeyPress={this.handleCommentKeyPress}
            value={newComment}
          />
        </form>
      </div>
    );
  }

  render() {
    const { post } = this.props;
    return (
      <div className="post-block">
        {this.renderMenu()}
        <h5 className="post-title"><em>{post.title}</em></h5>
        <ReactMarkdown className="post-content" plugins={[breaks]} source={post.content} />
        {this.renderCommentSection()}
      </div>
    );
  }
}

Post.propTypes = {
  post: PropTypes.shape({
    id: PropTypes.string.isRequired,
    source: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    authorId: PropTypes.string.isRequired,
    published: PropTypes.string.isRequired,
    title: PropTypes.string,
    content: PropTypes.string,
    comments: PropTypes.array,
    isGithubPost: PropTypes.bool,
  }).isRequired,
  invisible: PropTypes.bool,
  previewMode: PropTypes.bool,
  onEdit: PropTypes.func,
  onDelete: PropTypes.func,
  onNewComment: PropTypes.func,
};

Post.defaultProps = {
  invisible: false,
  previewMode: false,
  onEdit: null,
  onDelete: null,
  onNewComment: null,
};

export default Post;
