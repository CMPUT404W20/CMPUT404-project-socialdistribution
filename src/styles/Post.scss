$BACKGROUND-WHITE: #ffffff;
$left: 25px;
$right: 25px;
$top: 15px;
$bottom: 15px;
$post-width: 500px;

.post-block {
  width: $post-width;
  height: fit-content;
  background-color: $BACKGROUND-WHITE;
  box-shadow: 0px 0px 6px rgba(202, 202, 202, 0.25);
}
.post-username,
.post-time {
  display: inline;
}

.post-info {
  padding: $top $right $bottom $left;
  margin-bottom: $bottom+10px;
}

.post-username {
  text-align: left;
  font-weight: bold;
  float: left;
}
.post-time {
  text-align: right;
  font-weight: lighter;
  margin-right: 0vw;
  float: right;
}
.post-img {
  width: 100%;
}
.post-content {
  text-align: left;
  padding: $top $right $bottom $left;
}
#post-more-icon {
  float: right;
  margin-left: 10px;
}
