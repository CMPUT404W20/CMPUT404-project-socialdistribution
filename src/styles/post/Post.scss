@import "../common.scss";
.post-block {
    width: 100%;
    height: fit-content;
    background-color: $post-background-color;
    box-shadow: $post-box-shadow;
    border-radius: $radius;
    overflow: auto;

    .post-info {
        padding: 0 15pt;
        height: 30pt;
        font-size: 12px;
        .post-user-and-visibility {
            color: #aaaaaa;
            text-align: left;
            float: left;
            vertical-align: middle;
            position: relative;
            top: 0.5pt;
            line-height: 30pt;
            font-size: 14px;
            svg {
                position: relative;
                top: 1.5pt;
                left: 2pt;
                height: 100%;
            }

            .github-icon {
                width: 20px;
                margin-bottom: 3pt;
                margin-right: 10pt;
                opacity: 30%;
            }
        }
        .post-time {
            color: #aaaaaa;
            float: right;
            line-height: 30pt;
        }
        #post-more-button {
            all: unset;
            border: none;
            outline: none;
            float: right;
            margin-left: 10px;
            line-height: 30pt;
        }
        .dropdown-menu {
            @extend %dropdown;
        }
        .dropdown-toggle::after {
            border: none;
        }
    }
    .post-img {
        width: 100%;
    }
    .post-show-comment {
        all: unset;
        float: right;
        margin: 0 2rem 1rem 0;
        color: $icon-color;
        font-size: 11pt;
    }
    .comment-list {
        padding: 2rem 15pt 1rem 15pt;
        p {
            margin-bottom: 5pt;
            font-size: 10pt;
            color: $icon-color;
        }
        span {
            color: black;

            &.op {
                color: $accent-color;
            }
        }
    }
    .comment-content {
        padding-left: 0.5rem;
    }
    .post-comment-text-area {
        width: 100%;
        min-height: 40px;
        font-size: 10pt;
        padding: 10pt 15pt 5pt 15pt;
        border: none;
        outline: none;
        border-top: 0.4px solid #E5E5E5;
        resize: none;
    }
    .post-title {
        margin-top: -10pt;
        padding: 10pt 15pt 0 15pt;
        letter-spacing: -0.05em;
    }
    .post-content {
        text-align: left;
        padding: 0px 15pt 5px 15pt;
        line-height: normal;
        // markdown styles below
        h1 {
            font-size: 20pt;
        }
        h2 {
            font-size: 18pt;
        }
        h3 {
            font-size: 16pt
        }
        h4 {
            font-size: 14pt;
        }
        h5 {
            font-size: 12pt;
        }
        h6 {
            font-size: 10pt;
        }
        pre {
            border: 1px solid black;
            padding: 10px;
            border-radius: $radius;
            background-color: $background-color;
        }
        blockquote {
            border-left: 3px solid black;
            margin-top: 14px;
            padding: 2px 0px 2px 10px;
            p {
                margin-top: 1rem;
            }
        }
    }
    .spacer {
        height: 15pt;
        width: 100%;
    }
}