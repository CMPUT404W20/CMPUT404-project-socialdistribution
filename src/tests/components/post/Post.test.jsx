import React from "react";
import { shallow } from "enzyme";
import Post from "../../../components/post/Post";
import demoImage from "../../../images/demo-img.png";

const post = {
  id: 1,
  username: "username",
  imageSrc: demoImage,
  postTime: new Date(),
  content: "content",
};

describe("Post Block Components", () => {
  it("should render correctly", () => {
    const component = shallow(
      <Post
        post={post}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it("should render markdown content correctly", () => {
    const mdContent = `# Heading 1
## Heading 2
### Heading 3
#### Heading 4
##### Heading 5
###### Heading 6

1. Bullet 1
2. Bullet 2

[Google Link](https://www.google.com)

![alt text](https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png "Some logo")

Inline \`code\` 

\`\`\`
var s = "some code";
alert(s);
\`\`\`

| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| col 3 is      | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |

> Blockquote

---

Normal Text
    `;
    post.content = mdContent;
    const component = shallow(
      <Post
        post={post}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
