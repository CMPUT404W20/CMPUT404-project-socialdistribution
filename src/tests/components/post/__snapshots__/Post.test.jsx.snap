// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Block Components should render correctly 1`] = `
<div
  className="post-block"
>
  <div
    className="post-info"
  >
    <span
      className="post-user-and-visibility"
    >
      username
    </span>
    <DropdownButton
      alignRight={true}
      drop="down"
      id="post-more-button"
      title={
        <img
          alt="more-icon"
          id="post-more-icon"
          src="more-icon.svg"
        />
      }
    >
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </DropdownItem>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </DropdownItem>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        href="#"
      >
        Copy Link
      </DropdownItem>
    </DropdownButton>
    <div
      className="post-time"
    >
      a few seconds ago
    </div>
  </div>
  <img
    alt="more-icon"
    className="post-img"
    src="demo-img.png"
  />
  <ReactMarkdown
    astPlugins={Array []}
    className="post-content"
    escapeHtml={true}
    parserOptions={Object {}}
    plugins={
      Array [
        [Function],
      ]
    }
    rawSourcePos={false}
    renderers={Object {}}
    skipHtml={false}
    source="content"
    sourcePos={false}
    transformLinkUri={[Function]}
  />
  <div>
    <button
      aria-controls="post-comments"
      aria-expanded={false}
      className="post-show-comment"
      onClick={[Function]}
      type="button"
    >
      1
       
      comment
    </button>
    <Collapse
      appear={false}
      dimension="height"
      getDimensionValue={[Function]}
      in={false}
      mountOnEnter={false}
      timeout={300}
      unmountOnExit={false}
    >
      <div>
        <div
          id="comment-list"
        >
          <div
            key="95c76f63-5579-45f6-810f-e0665f2ea588"
          >
            <p>
              commentAuthor:
              <span
                className="comment-content"
              >
                Sample Comment
              </span>
            </p>
          </div>
        </div>
      </div>
    </Collapse>
    <form
      action="submit"
      className="make-comment-input-wrapper"
      onSubmit={[Function]}
    >
      <TextareaAutosize
        className="post-comment-text-area"
        inputRef={[Function]}
        onChange={[Function]}
        onHeightChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Add a comment"
        useCacheForDOMMeasurements={false}
        value=""
      />
    </form>
  </div>
</div>
`;

exports[`Post Block Components should render markdown content correctly 1`] = `
<div
  className="post-block"
>
  <div
    className="post-info"
  >
    <span
      className="post-user-and-visibility"
    >
      username
    </span>
    <DropdownButton
      alignRight={true}
      drop="down"
      id="post-more-button"
      title={
        <img
          alt="more-icon"
          id="post-more-icon"
          src="more-icon.svg"
        />
      }
    >
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        onClick={[Function]}
      >
        Edit
      </DropdownItem>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </DropdownItem>
      <DropdownItem
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "SafeAnchor",
            "render": [Function],
          }
        }
        disabled={false}
        href="#"
      >
        Copy Link
      </DropdownItem>
    </DropdownButton>
    <div
      className="post-time"
    >
      a few seconds ago
    </div>
  </div>
  <img
    alt="more-icon"
    className="post-img"
    src="demo-img.png"
  />
  <ReactMarkdown
    astPlugins={Array []}
    className="post-content"
    escapeHtml={true}
    parserOptions={Object {}}
    plugins={
      Array [
        [Function],
      ]
    }
    rawSourcePos={false}
    renderers={Object {}}
    skipHtml={false}
    source="# Heading 1
## Heading 2
### Heading 3
#### Heading 4
##### Heading 5
###### Heading 6

1. Bullet 1
2. Bullet 2

[Google Link](https://www.google.com)

![alt text](https://github.com/adam-p/markdown-here/raw/master/src/common/images/icon48.png \\"Some logo\\")

Inline \`code\` 

\`\`\`
var s = \\"some code\\";
alert(s);
\`\`\`

| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| col 3 is      | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |

> Blockquote

---

Normal Text
    "
    sourcePos={false}
    transformLinkUri={[Function]}
  />
  <div>
    <button
      aria-controls="post-comments"
      aria-expanded={false}
      className="post-show-comment"
      onClick={[Function]}
      type="button"
    >
      1
       
      comment
    </button>
    <Collapse
      appear={false}
      dimension="height"
      getDimensionValue={[Function]}
      in={false}
      mountOnEnter={false}
      timeout={300}
      unmountOnExit={false}
    >
      <div>
        <div
          id="comment-list"
        >
          <div
            key="95c76f63-5579-45f6-810f-e0665f2ea588"
          >
            <p>
              commentAuthor:
              <span
                className="comment-content"
              >
                Sample Comment
              </span>
            </p>
          </div>
        </div>
      </div>
    </Collapse>
    <form
      action="submit"
      className="make-comment-input-wrapper"
      onSubmit={[Function]}
    >
      <TextareaAutosize
        className="post-comment-text-area"
        inputRef={[Function]}
        onChange={[Function]}
        onHeightChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Add a comment"
        useCacheForDOMMeasurements={false}
        value=""
      />
    </form>
  </div>
</div>
`;
